doc/final-report/final-report.md

```diff
@@ -0,0 +1,1478 @@
+being specified for PCIe 2.0, while in practice the RPi5 is capable of PCI 3.0,
+which increases the available bandwidth.
+
+The i7i.xlarge and i8g.xlarge AWS EC2 instances exhibit somewhat strange
+behaviour:
+
+Their IOPS scores are much higher than advertised.
+
+: This can be explained by the fact that these machines are virtual and use
+  shared hardware, so that one can presume that their rated IOPS are guaranteed
+  minimums that can be exceeded when the physical machines’ IOPS are
+  under-utilised.
+
+Their I/O is bursty.
+
+: The bursty behaviour is that initial `fio` runs achieve even higher IOPS
+  scores (500 k and 450 k, respectively, on one core), but after several runs
+  the scores settle down to the numbers reported in the table above. This
+  behaviour is documented by AWS itself: it is due to disk I/O ‘credits’ that
+  allow short bursts at higher performance, which cannot be sustained.
+
+The IOPS scores scale negatively when adding more cores.
+
+: This is actually quite bizarre: the more cores used to submit I/O, the lower
+  the aggregate IOPS that is measured. Apparently, this is not the result of a
+  measurement artefact but shows a real effect, and it is *opposite* to what
+  happens with physical hardware. Running `fio` on the i8g.xlarge machine with
+  4 cores results in 175 k IOPS (which is near to the rated 150 k IOPS), showing
+  that the negative scaling continues beyond two cores.
+
+The IOPS scores of i7i.xlarge and i8g.xlarge are the same.
+
+: This can be explained by the fact that both machines use the same generation
+  of the Nitro hypervisor and I/O hardware.
+
+These behaviours are worth keeping in mind for deployment. In particular, a
+benchmark may give too rosy a picture if the physical machine was not saturated
+at the time the benchmark was run, or the benchmark was not run long enough, or
+the benchmark does not use the same amount of I/O parallelism as the deployment.
+
+Finally, the dev laptop’s scores of 262 k IOPS on 1 core and 487 k on 2 cores
+are well below the laptop’s rated 1.4 M IOPS. The explanation is that the 1.4 M
+is for the SSD device as a whole and it takes multiple cores to saturate the
+SSD, while the benchmark numbers above are for a single core. On the same
+machine, the scores of the `fio` and `blockio-uring` benchmarks for 8 cores are
+1,449 k and 1,337 k IOPS respectively.
+
+## Setup of the primary benchmark
+
+The primary benchmark handles the required UTxO-like workload. Its starting
+point is a table with 100 million key–value pairs, with keys and values having
+the required sizes and keys being uniformly distributed. The benchmark reads
+this table from a snapshot and executes a sequence of batches of operations on
+it. Each batch consists of 256 lookups, 256 inserts and 256 deletes. To achieve
+reasonable running times, we use 10 k batches of operations for the serial mode
+and 100 k batches for the pipelined mode.
+
+### Serial and pipelined execution
```

Comment:
If I were reading this for the first time without prior knowledge about pipelining, I think I would understand the general idea of overlapping operations, but I fear I wouldn't fully understand the design as it is realised in our benchmark because the current description is still rather high-level. Maybe we can direct readers to the `utxo-db-api` document, which has a more formal explanation of pipelining and the dataflow, that might be helpful? Or maybe it's fine if the user has only a general intuition about pipelining, but then it might be helpful still to refer to the more formal description in the `utxo-db-api` doc

Reply:
> These constraints leave room for concurrent and parallel execution, because they
allow overlapped execution of multiple batches in a pipelined way. The reason
why such an execution is possible is somewhat subtle though. We provide a high
level summary here. For a full formal treatment see our previous work
[@utxo-db-api, Sections 7, 7.5].